sys/fw_slots/fw_slots.c

```diff
@@ -0,0 +1,477 @@
+                oldest_firmware_version = fw_slot_metadata.version;
+            }
+        } else {
+            oldest_fw_slot = slot;
+            oldest_firmware_version = fw_slot_metadata.version;
+        }
+    }
+
+    printf("[fw_slots] Oldest FW slot: #%d; Firmware v%d\n", oldest_fw_slot,
+            oldest_firmware_version);
+
+    return oldest_fw_slot;
+}
+
+int fw_slots_find_newest_int_image(void)
+{
+    /* At first, we only assume knowledge of version v0 */
+    int newest_fw_slot = 0;
+    uint16_t newest_firmware_version = 0;
+
+    /* Iterate through each of the MAX_FW_SLOTS. */
+    for (int slot = 1; slot <= MAX_FW_SLOTS ; slot++) {
+        /* Get the metadata of the current FW slot. */
+        FW_metadata_t fw_slot_metadata;
+
+        if(fw_slots_get_int_slot_metadata(slot, &fw_slot_metadata) == 0) {
+            fw_slots_print_metadata(&fw_slot_metadata);
+        } else {
+            printf("[fw_slots] ERROR cannot get slot metadata.\n");
+        }
+
+        /* Is this slot populated? If not, skip. */
+        if (fw_slots_validate_metadata(&fw_slot_metadata) == -1) {
+            continue;
+        }
+
+        /* Is this the newest non-Golden Image image we've found thus far? */
+        if ( fw_slot_metadata.version > newest_firmware_version ) {
+            newest_fw_slot = slot;
+            newest_firmware_version = fw_slot_metadata.version;
+        }
+    }
+
+    printf("Newest FW slot: #%d; Firmware v%d\n", newest_fw_slot,
+            newest_firmware_version);
+
+    return newest_fw_slot;
+}
+
+int fw_slots_erase_int_image(uint8_t fw_slot)
+{
+    /* Get page address of the fw_slot in internal flash */
+    uint32_t fw_image_base_address;
+    /* Get the page where the fw_slot is located */
+    uint8_t slot_page;
+
+    if (fw_slot > MAX_FW_SLOTS || fw_slot == 0) {
+        printf("[fw_slots] FW slot not valid, should be <= %d and > 0\n",
+                MAX_FW_SLOTS);
+        return -1;
+    }
+
+    fw_image_base_address = fw_slots_get_slot_address(fw_slot);
+
+    printf("[fw_slots] Erasing FW slot %u [%#lx, %#lx]...\n", fw_slot,
+            fw_image_base_address,
+            fw_image_base_address + (FW_SLOT_PAGES * FLASHPAGE_SIZE) - 1);
+
+    slot_page = fw_slots_get_slot_page(fw_slot);
+
+    /* Erase each page in the FW internal slot! */
+    for (int page = slot_page; page < slot_page + FW_SLOT_PAGES; page++) {
+        DEBUG("[fw_slots] Erasing page %d\n", page);
+        flashpage_write(page, NULL);
+    }
+
+    printf("[fw_slots] Erase successful\n");
+
+    return 0;
+}
+
+int fw_slots_update_firmware(uint8_t fw_slot)
+{
+    /*
+     * TODO
+     */
+    return 0;
+}
+
+int fw_slots_store_fw_data( uint32_t ext_address, uint8_t *data, size_t data_length )
+{
+    /*
+     * TODO
+     */
+    return 0;
+}
+
+/*
+ * _estack pointer needed to reset PSP position
+ */
+extern uint32_t _estack;
+
+void fw_slots_jump_to_image(uint32_t destination_address)
```

Comment:
This should be implemented per architecture! Maybe add in `cpu/stm32f1/cpu.c` something called `cpu_branch_address`.